File: src/main/java/com/rtomyj/yugiohAPI/model/BanListRemovedContent.java
package com.rtomyj.yugiohAPI.model;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.rtomyj.yugiohAPI.controller.banlist.CardsController;
import com.rtomyj.yugiohAPI.controller.banlist.NewController;
import com.rtomyj.yugiohAPI.controller.banlist.RemovedController;

import lombok.*;
import org.springframework.hateoas.RepresentationModel;

import io.swagger.annotations.ApiModel;

@Data
@EqualsAndHashCode(callSuper=false)
@Builder
@NoArgsConstructor
@AllArgsConstructor
@ApiModel(description = "Information about a ban list (listRequested) obtained by comparing to a previous ban list (comparedTo). The information retrieved will be the removedCards.")
@JsonPropertyOrder({ "listRequested", "comparedTo", "numRemoved", "removedCards" })
public class BanListRemovedContent extends RepresentationModel<BanListRemovedContent>
{
	private String listRequested;
	private String comparedTo;

	private int numRemoved;
	private List<BanListComparisonResults> removedCards;

	private static final Class<RemovedController> controllerClass = RemovedController.class;
	private static final Class<CardsController> banListController = CardsController.class;
	private static final Class<NewController> newController = NewController.class;



	private void setLink()
	{
		this.add(
			linkTo(methodOn(controllerClass).getRemovedContent(listRequested)).withSelfRel()
		);

		this.add(
			linkTo(methodOn(banListController).getBannedCards(listRequested, false)).withRel("banlist")
		);

		this.add(
			linkTo(methodOn(newController).getNewContent(listRequested)).withRel("new-cards")
		);

		this.add(
			linkTo(methodOn(controllerClass).getRemovedContent(comparedTo)).withRel("previous")
		);
	}



	public void setLinks()
	{
		this.setLink();
		BanListComparisonResults.setLinks(removedCards);
	}
}

File: src/main/kotlin/com/rtomyj/skc/find/banlist/model/BanListRemovedContent.kt
package com.rtomyj.skc.banlist.model

import com.fasterxml.jackson.annotation.JsonPropertyOrder
import com.rtomyj.skc.Open
import com.rtomyj.skc.banlist.controller.BanListDiffController
import com.rtomyj.skc.banlist.controller.BannedCardsController
import com.rtomyj.skc.constant.SwaggerConstants
import com.rtomyj.skc.util.HateoasLinks
import io.swagger.v3.oas.annotations.media.Schema
import org.springframework.hateoas.RepresentationModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder

@JsonPropertyOrder("listRequested", "comparedTo", "numRemoved", "removedCards")
@Schema(
    implementation = BanListRemovedContent::class,
    description = "Cards that were removed from a ban list compared to the previous logical ban list.",
)
@Open
data class BanListRemovedContent(
	@Schema(
        implementation = String::class,
        description = SwaggerConstants.BAN_LIST_START_DATE_DESCRIPTION,
    )
    val listRequested: String,

	@Schema(
        implementation = String::class,
        description = SwaggerConstants.PREVIOUS_BAN_LIST_START_DATE_DESCRIPTION,
    )
    val comparedTo: String,

	@Schema(
        implementation = List::class,
        description = "List containing removed cards and their previous ban status.",
    )
    val removedCards: List<CardsPreviousBanListStatus>,

	@Schema(
        implementation = Int::class,
        description = "Total number of cards removed in requested ban list compared to previous ban list.",
    )
    val numRemoved: Int = removedCards.size
) : RepresentationModel<BanListRemovedContent?>(), HateoasLinks {

    companion object {
        private val banListController = BannedCardsController::class.java
        private val diffController = BanListDiffController::class.java
    }


    override fun setSelfLink() {
        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(diffController).getNewlyRemovedContentForBanList(listRequested)
            ).withSelfRel()
        )
    }

    override fun setLinks() {
        setSelfLink()

        removedCards
            .forEach(CardsPreviousBanListStatus::setLinks)

        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(banListController).getBannedCards(listRequested, false, true)
            ).withRel("Ban List Content")
        )
        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(diffController).getNewlyAddedContentForBanList(listRequested)
            ).withRel("Ban List New Content")
        )
    }
}